text/0000-path-clarity.md

```diff
@@ -0,0 +1,734 @@
+mod submodule {
+    use futures;
+    fn my_poll() -> futures::Poll { ... }
+}
+```
+
+Together with the fact that you use `crate::` in `use` declarations, this
+strongly reinforces the idea that:
+
+- `use` brings items into scope, based on paths that start by identifying the crate
+- an item needs to be in scope before you can refer to it
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+First, a bit of terminology: a *fully qualified path* is a path starting with
+`::`, which *all* paths in `use` do implicitly.
+
+The actual changes in this RFC are fairly small tweaks to the current module
+system; most of the complexity comes from the migration plans.
+
+There are two basic ways we could do migration. The first one, pioneered by [RFC
+2088], treats external crates as a kind of "fallback" when interpeting fully
+qualified paths (to retain compatibility with today's code). That solution is
+detailed in the alternatives section.
+
+[RFC 2088]: https://github.com/rust-lang/rfcs/pull/2088
+
+This RFC, however, spells out a different alternative which avoids the need for
+any kind of fallback in name resolution, or indeed any "implicit" linking of
+crates at all. This is achieved via a two-stage approach: one within the current
+[epoch], and one for the next.
+
+[epoch]: https://github.com/rust-lang/rfcs/pull/2052
+
+## Stage 1 (current epoch)
+
+The core idea of stage 1 is to make all of the changes in this RFC *except* for
+removing the need for `extern crate`. In particular, this allows people to
+migrate paths to use `crate::` in preparation for the next stage.
+
+In detail, at stage 1:
+
+- You can write `mod bar;` statements even when not in a `mod.rs` or equivalent;
+  in this case, the submodules must appear within a subdirectory with
+  the same name as the current module. Thus, `foo.rs` can contain `mod bar;` if
+  there is also a `foo/bar.rs`.
+  - It is not permitted to have both `foo.rs` and `foo/mod.rs` at the same point
+    in the file system.
+
+- We introduce `crate` as a new visibility specifier, shorthand for `pub(crate)`
+  visibility.
+
+- We introduce a lint against use of bare `pub` for items which are not
+  reachable via some fully-`pub` path. That is, bare `pub` should truly mean
+  *public*, and `crate` should be used for crate-level visibility.
+
+- We introduce `crate` as a new path component which designates the root of the
+  current crate, and deprecate fully qualified paths that do not begin with one
+  of: an external crate name, `crate`, `super`, or `self`.
+  - However, at this stage, external crate names *must* be provided by
+    *explicit* `extern crate` declarations at the root of the crate.
+  - We deprecate any use of `extern crate` *not* at crate root.
+
+- Cargo will provide a new `alias` key for dependencies, so that e.g. users who
+  want to use the `rand` crate but call it `random` instead can now write `rand
+  = { version = "0.3", alias = "random" }`.
```

Comment:
Perhaps I'm missing something obvious, but how does rustc get this `alias` information? Does it get passed by cargo or does rustc parse the Cargo.toml? The RFC says that the interface between rustc and cargo doesn't change, so I'm a bit confused here...

Reply:
Cargo already passes `--extern rand=path/to/rand.rlib`, all it has to do is change the first `rand` to be `random`, but the interface remains the same. Maybe the RFC should be more explicit here?

Reply:
It's not spelled out here, but in previous proposals, the idea was for cargo to pass `--extern foo=.../libbar.rlib`.  This form of aliasing is already supported by rustc, so the only change is to let cargo make use of it.

Reply:
Ah, that makes sense. Perhaps this could be added to the RFC, please, @aturon ?